src/pytesmo/validation_framework/metric_calculators_adapters.py

```diff
@@ -0,0 +1,159 @@
+# ARE DISCLAIMED. IN NO EVENT SHALL VIENNA UNIVERSITY OF TECHNOLOGY,
+# DEPARTMENT OF GEODESY AND GEOINFORMATION BE LIABLE FOR ANY
+# DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+# THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+
+"""
+Metric Calculator Adapters change how metrics are calculated by calling
+the `calc_metric` function of the adapted calculator instead of the unadapted
+version.
+"""
+
+from pytesmo.validation_framework.metric_calculators import (
+    PairwiseIntercomparisonMetrics,
+    TripleCollocationMetrics
+)
+import warnings
+import numpy as np
+
+
+class MonthsMetricsAdapter(object):
+    """
+    Adapt MetricCalculators to calculate metrics for groups across months
+    """
+
+    _supported = (
+        PairwiseIntercomparisonMetrics,
+        TripleCollocationMetrics,
+    )
+
+    def __init__(self, calculator, sets=None):
+        """
+        Add functionality to a metric calculator to calculate validation
+        metrics for subsets of certain months in a time series (e.g. seasonal).
+
+        Parameters
+        ----------
+        calculator : MetadataMetrics or any child of it (not Rolling metrics)
+        sets : dict, optional (default: None)
+            A dictionary consisting of a set name (which is added to the metric
+            name as a suffix) and the list of months that belong to that set.
+            If None is passed, we use 4 (seasonal) sets named after the fist
+            letter of each month used.
+        """
+        if not isinstance(calculator, self._supported):
+            warnings.warn(f"Adapting {calculator.__class__} is not supported.")
+        self.cls = calculator
+        if sets is None:
+            sets = {
+                "DJF": [12, 1, 2],
+                "MAM": [3, 4, 5],
+                "JJA": [6, 7, 8],
+                "SON": [9, 10, 11],
+                "ALL": list(range(1, 13)),
+            }
+
+        self.sets = sets
+
+        # metadata metrics and lon, lat, gpi are excluded from applying
+        # seasonally
+        self.non_seas_metrics = ["gpi", "lon", "lat"]
+        if self.cls.metadata_template is not None:
+            self.non_seas_metrics += list(self.cls.metadata_template.keys())
+
+        all_metrics = calculator.result_template
+        subset_metrics = {}
+
+        # for each subset create a copy of the metric template
+        for name in sets.keys():
+            for k, v in all_metrics.items():
+                if k in self.non_seas_metrics:
+                    subset_metrics[f"{k}"] = v
+                else:
+                    subset_metrics[(f"{name}", f"{k}")] = v
```

Comment:
Are you testing that this is really working as expected? I think that this is only doing shallow copies, not duplicating the metric templates. These should normally be copied inside the metric calculator, but it could lead to unexpected behaviour if someone doesn't know about this. Since `v` are arrays, you could just replace `v` with `np.array(v)`.